ipq807x: add support for Linksys MX4200 V1 and V2

SpectreDev1 opened the pull request:
Linksys MX4200 is a 802.11ax Tri-band router/AP.
Specifications:
* CPU: Qualcomm IPQ8174 Quad core Cortex-A53 1.4GHz
* RAM: 512MB of DDR3
* RAM: 1GB of DDR3 For V2
* Storage: 512Mb NAND
* Ethernet: 4x1G RJ45 ports (QCA8075)
* Bluetooth: EFR32MG21
* WLAN:
	* 2.4GHz: Qualcomm QCN5024 2x2 802.11b/g/n/ax 574 Mbps PHY rate
	* 5GHz: Qualcomm QCN5054 2x2@80MHz  802.11a/b/g/n/ac/ax 2402 PHY rate
	* 5GHz: Qualcomm QCN5054 4x4@80MHz 802.11a/b/g/n/ac/ax 2402 PHY rate
* LED-s:
	* RGB system led

* Buttons: 1x Soft reset 1x WPS
* Power: 12V DC Jack

Installation instructions:
Open Linksys Web UI - http://192.168.1.1/ca or http://10.65.1.1/ca depending on your setup.
Login with your admin password. The default password can be found on a sticker under the device.
To enter into the support mode, click on the “CA” link and the bottom of the page.
Open the “Connectivity” menu and upload the squash-factory image with the “Choose file” button.
Click start. Ignore all the prompts and warnings by click “yes” in all the popups.
The Wifi radios are turned off by default. To configure the router, you will need to connect your computer to the LAN port of the device.
Then you would need to write openwrt to the other partition for it to work
- First Check booted partition
fw_printenv -n boot_part

- Then install Openwrt to the other partition if booted in slot 1:
mtd -r -e alt_kernel -n write openwrt-qualcommax-ipq807x-linksys_mx4200v(X)-squashfs-factory.bin alt_kernel

- If in slot 2:
mtd -r -e kernel -n write openwrt-qualcommax-ipq807x-linksys_mx4200v(X)-squashfs-factory.bin kernel

Replace (X) with your model version either 1 or 2

robimarko commented on package/firmware/ipq-wifi/Makefile:
```diff
@@ -114,14 +114,16 @@ Do not install it for any other device!
 $(eval $(call generate-ipq-wifi-package,edgecore_eap102,Edgecore EAP102))
 $(eval $(call generate-ipq-wifi-package,edimax_cax1800,Edimax CAX1800))
+$(eval $(call generate-ipq-wifi-package,linksys_mx4200v1,Linksys MX4200))
+$(eval $(call generate-ipq-wifi-package,linksys_mx4200v2,Linksys MX4200))
```
Do these BDFs differ at all?

SpectreDev1 replied:
I haven't actually tested that, but I don't think they differ

robimarko replied:
Wait, where are these even coming from?
I am not seeing them in the BDF repo, nor I am seeing a PR for them.

Please binary diff the stock BDF-s first, and then package them.

SpectreDev1 replied:
yes i was about to create a pull request add the BDFs

SpectreDev1 replied:
done

robimarko replied:
But you are still not even generating the meta package.

You need to add your device to:
https://github.com/openwrt/openwrt/blob/main/package/firmware/ipq-wifi/Makefile#L30

SpectreDev1 replied:
some how i missed that, ive added it now

robimarko commented on package/boot/uboot-envtools/files/qualcommax_ipq807x:
```diff
@@ -27,6 +27,13 @@ edimax,cax1800)
+linksys,mx4200v2)
+	idx="$(find_mtd_index u_env)"
+	[ -n "$idx" ] && \
+		ubootenv_add_uci_config "/dev/mtd$idx" "0x0" "0x40000" "0x20000"
```
You are missing the number of NAND blocks here

SpectreDev1 replied:
Do you mean the erase size?

robimarko replied:
No, I mean how many blocks on the NAND does the environment take?
Considering that you are saying the size is `0x40000` and your NAND has block size of 128k that means 2 blocks.

SpectreDev1 replied:
> No, I mean how many blocks on the NAND does the environment take? Considering that you are saying the size is `0x40000` and your NAND has block size of 128k that means 2 blocks.

yes it takes up 0x40000, this has been calculated from the stock log

robimarko replied:
Why has this been marked as solved?

The number of NAND blocks has not been added to config.

robimarko replied:
Quite simply, you have the MTD partition, total size, and block size set but no the umber of blocks.

So, add `"2"` at the end of `ubootenv_add_uci_config`

SpectreDev1 replied:
Done

robimarko commented on target/linux/qualcommax/image/ipq807x.mk:
```diff
@@ -91,6 +91,38 @@ define Device/edimax_cax1800
+endef
+TARGET_DEVICES += linksys_mx4200v1
+
+define Device/linksys_mx4200v2
```
Please reuse the recipe from v1 or make a common recipe instead of duplicating everything as these barely differ.

robimarko commented on target/linux/qualcommax/image/ipq807x.mk:
```diff
@@ -91,6 +91,38 @@ define Device/edimax_cax1800
+	UBINIZE_OPTS := -E 5    # EOD marks to "hide" factory sig at EOF
+	IMAGES += factory.bin
+	IMAGE/factory.bin  := append-kernel | pad-to $$$$(KERNEL_SIZE) | append-ubi | linksys-image type=MX4200
+	DEVICE_PACKAGES := kmod-leds-pca963x ipq-wifi-linksys_mx4200 kmod-usb-storage-uas kmod-usb-storage-extras kmod-usb-dwc2-pci kmod-usb-uhci kmod-fs-ext
```
Remove USB and EXT4 kmods, those are option for the device to function and can be easily installed.

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8174-mx4200.dtsi:
```diff
@@ -0,0 +1,435 @@
+
+			partition@1080000 {
+				label = "kernel";
+				reg = <0x1080000 0x9600000>;
```
This is obviously completely incorrect, you are claiming that kernel size is 150MB?

SpectreDev1 replied:
This is how its done on stock, so ive taken it from there

SpectreDev1 replied:
in the logs the partition match stock fw fully

robimarko replied:
But this doesn't make any sense, how can you have your kernel and roots partitions overlap?

SpectreDev1 replied:
yeah i dont know but this is how Linksys has done it, so ive just followed that, i can provide stock log if need be

robimarko replied:
Regardless of stock FW doing it, it just doesn't make any sense nor I see how it could work.
Does it even have dual FW or are they just faking it?

SpectreDev1 replied:
it has dual fw as far as i know

SpectreDev1 commented on package/firmware/ipq-wifi/Makefile:
Pull Request to add board file here https://github.com/openwrt/firmware_qca-wireless/pull/13

robimarko commented on target/linux/qualcommax/image/ipq807x.mk:
```diff
@@ -91,6 +91,33 @@ define Device/edimax_cax1800
+endef
+TARGET_DEVICES += linksys_mx4200v2
+
+define Device/linksys_mx4200
```
This goes before the device profiles.

SpectreDev1 replied:
Done

robimarko commented on target/linux/qualcommax/image/ipq807x.mk:
```diff
@@ -91,6 +91,33 @@ define Device/edimax_cax1800
 
+define Device/linksys_mx4200v1
+	$(call Device/linksys_mx4200)
+	DEVICE_MODEL := MX4200v1
```
Use `DEVICE_VARIANT` to set v1 and v2.

SpectreDev1 replied:
i think ive solved that now

robimarko commented on target/linux/qualcommax/image/ipq807x.mk:
```diff
@@ -91,6 +91,33 @@ define Device/edimax_cax1800
+define Device/linksys_mx4200v1
+	$(call Device/linksys_mx4200)
+	DEVICE_MODEL := MX4200v1
+	DEVICE_PACKAGES := kmod-ath10k-ct-smallbuffers
```
Does v1 really have an ath10k compatible radio?

robimarko commented on package/boot/uboot-envtools/files/qualcommax_ipq807x:
```diff
@@ -27,6 +27,13 @@ edimax,cax1800)
+linksys,mx4200v2)
+	idx="$(find_mtd_index u_env)"
+	[ -n "$idx" ] && \
+		ubootenv_add_uci_config "2" "/dev/mtd$idx" "0x0" "0x40000" "0x20000"
```
2 goes at the end, after 0x20000

robimarko commented on target/linux/qualcommax/image/ipq807x.mk:
```diff
@@ -91,6 +91,34 @@ define Device/edimax_cax1800
+
+define Device/linksys_mx4200v1
+	$(call Device/linksys_mx4200)
+	DEVICE_MODEL := MX4200
```
Move it to the common profile, no need to duplicate in each version

SpectreDev1 commented:
i have one issue with this and i cant seem to find the solution for it, if i do sysupgrade from alt_rootfs it switches slots and tries to boot rootfs however it just stops at a point because its `waiting for root device /dev/ubiblock0_1`

judahrand commented:
> i have one issue with this and i cant seem to find the solution for it, if i do sysupgrade from alt_rootfs it switches slots and tries to boot rootfs however it just stops at a point because its `waiting for root device /dev/ubiblock0_1`

I have literally no idea what I'm talking about and unfortunately I don't have a spare MX4200 to test with but have you tried adding `ubi.block=0,rootfs` to the `bootargs-append`? 

I've been trying to find the logs in the various pastebins posted on the forum but they're all expired would be great to get a full set up logs posted up here.

SpectreDev1 commented:
> > i have one issue with this and i cant seem to find the solution for it, if i do sysupgrade from alt_rootfs it switches slots and tries to boot rootfs however it just stops at a point because its `waiting for root device /dev/ubiblock0_1`
> 
> I have literally no idea what I'm talking about and unfortunately I don't have a spare MX4200 to test with but have you tried adding `ubi.block=0,rootfs` to the `bootargs-append`? I see it used on the Netgear WAX218 and elsewhere.
> 
> I've been trying to find the logs in the various pastebins posted on the forum but they're all expired would be great to get a full set up logs posted up here.

I don't need to since its already done by stock

SpectreDev1 commented:
> > i have one issue with this and i cant seem to find the solution for it, if i do sysupgrade from alt_rootfs it switches slots and tries to boot rootfs however it just stops at a point because its `waiting for root device /dev/ubiblock0_1`
> 
> I have literally no idea what I'm talking about and unfortunately I don't have a spare MX4200 to test with but have you tried adding `ubi.block=0,rootfs` to the `bootargs-append`? I see it used on the Netgear WAX218 and elsewhere.
> 
> I've been trying to find the logs in the various pastebins posted on the forum but they're all expired would be great to get a full set up logs posted up here.

I'll send a log in a bit

judahrand commented:
I wonder if something in the logic implemented for `ipq40xx` isn't working as expected here and we're leaving behind some OEM volumes? That would suggest the the flash itself is what is failing rather than the boot?

https://github.com/openwrt/openwrt/commit/1440175f82f36c64de8a8593f68749741762580a

judahrand commented:
An OEM bootlog might be useful to help determine if that is the case?

SpectreDev1 commented:
@judahrand you got anywhere yet?

innovara commented:
I'm sorry if this message creates a bit of unnecessary noise but I just wanted to say thank you for all the hard work that you are doing to bring OpenWrt to the Linksys MX4200. I have a V2 unit and I am following attentively your efforts.

i-srinivas commented:
Thank you

> I'm sorry if this message creates a bit of unnecessary noise but I just wanted to say thank you for all the hard work that you are doing to bring OpenWrt to the Linksys MX4200. I have a V2 unit and I am following attentively your efforts.

Same here - though I have Netgear RBR750 with same chipset, and this PR provides insights into adding support for another device.

i-srinivas commented on package/boot/uboot-envtools/files/qualcommax_ipq807x:
```diff
@@ -27,6 +27,13 @@ edimax,cax1800)
+	[ -n "$idx" ] && \
+		ubootenv_add_uci_config "/dev/mtd$idx" "0x0" "0x40000" "0x20000" "2"
+	;;
+
```
IPQ8173/IPQ8174 is marketed as Qualcomm Pro 800 Networking platform, not sure if we want to combine with ipq807x

There are hardware differences between the 2, as provided by Qualcomm, though the cipset is the same 

https://www.qualcomm.com/products/internet-of-things/networking/wi-fi-networks/ipq8074
https://www.qualcomm.com/products/internet-of-things/networking/wi-fi-networks/networking-pro-series/qualcomm-networking-pro-800-platform

I would suggest to add a new platform onboarding for IPQ8174, rather than adding to ipq807x.

robimarko replied:
No, they are functionally the same when it comes to the software.
The main difference is CPU and NSS clocks, like with IPQ8070/1A.

SpectreDev1 replied:
yeah could be done later when theres other devices with this chipset added

robimarko replied:
There is really no point for Oak specific subtarget

SpectreDev1 replied:
> There is really no point for Oak specific subtarget

yeah after reading your comment, it'd be a waste of time todo it

judahrand commented:
> @judahrand you got anywhere yet?

I'm afraid not 😔 I had a bit of a tinker but have put it on hold for now.

SpectreDev1 commented:
> > @judahrand you got anywhere yet?
> 
> I'm afraid not 😔 I had a bit of a tinker but have put it on hold for now.

Hmm ok, quite an interesting issue, need to ask another linksys openwrt dev i guess

i-srinivas commented on package/firmware/ipq-wifi/Makefile:
```diff
@@ -76,7 +77,7 @@ define ipq-wifi-install-one
   $(if $(filter $(suffix $(1)),.QCA99X0 .qca99x0),\
     $(call ipq-wifi-install-one-to,$(1),$(2),QCA99X0/hw2.0),\
-  $(if $(filter $(suffix $(1)),.IPQ8074 .ipq8074),\
+  $(if $(filter $(suffix $(1)),.IPQ8074 .ipq8074 .ipq8174),\
```
I am following your PR changes and into forked repo. https://github.com/i-srinivas/openwrt/tree/rbr750 that I am working on spare time. 

Found the above and it seems - the code filters for spell cases. Do you think based on the above we also need to add upper case '.IPQ8174' here ?

SpectreDev1 replied:
Our file is in lower case, but yeah we could add the upper case name too

i-srinivas replied:
ok, lets keep this consistent - and helpful for newbies like me.

Ansuel replied:
Honestly for the sake of it I would keep ipq8074... @robimarko any problem with that?

i-srinivas replied:
FCC details shows IPQ8174, I do not know the intricacies between chip differences - it would be good to keep the chip # to avoid any confusion for people reading the code 

https://fccid.io/PY319400470/Internal-Photos/Internal-photos-4603355.pdf

Ansuel replied:
Those bdf extension are totally arbitrarily. (In fact if we should be precise there should be separation for ipq8072)
The extension is really to differentiate one family of BDF to another. The idea is to not be that precise about the bdf since they are all part of a bigger family.

SpectreDev1 commented:
Ok @Ansuel @robimarko sysupgrade is all working

robimarko commented:
Ok, squash the changes and I can do another round of reviews to get this ready

SpectreDev1 commented:
> Ok, squash the changes and I can do another round of reviews to get this ready

all done

robimarko commented:
Edit the commit description remove the leftover commit names after squashing.
After that is done:
`Reviewed-by: Robert Marko <robimarko@gmail.com>`

SpectreDev1 commented:
> Edit the commit description remove the leftover commit names after squashing. After that is done: `Reviewed-by: Robert Marko <robimarko@gmail.com>`

alright all done

innovara commented:
@SpectreDev1 I guess you are aware that there is a conflict on `package/firmware/ipq-wifi/Makefile` when merging into main now?

SpectreDev1 commented:
> @SpectreDev1 I guess you are aware that there is a conflict on `package/firmware/ipq-wifi/Makefile` when merging into main now?

Yeah I'll resolve it in a bit

SpectreDev1 commented:
conflict resolved

testuser7 commented:
You should correct information about first 5GHz radio. It's only 2x2@80MHz.

testuser7 commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8174-mx4200.dtsi:
```diff
@@ -0,0 +1,445 @@
+		ethernet4 = &dp5;
+		led-boot = &led_system_blue;
+		led-running = &led_system_blue;
+		led-failsafe = &led_system_red;
```
I think that for the failsafe we should use purple (blue+red).
OEM led description:
![mx4200_led](https://github.com/openwrt/openwrt/assets/10881063/55db6179-7339-4f58-a5b7-7472905ad204)
There is no color change during a firmware update.

SpectreDev1 replied:
Yeah but may aswell use green for update so you know it's updating but yeah we can use purple for failsafe

testuser7 replied:
Personally, I would prefer red when updating.

SpectreDev1 replied:
> Personally, I would prefer red when updating.

Yeah but green was never really used, may aswell make it for updating

testuser7 replied:
Or maybe cyan (blue+green)? It's not used either.

SpectreDev1 replied:
Yeah that can be done too

raenye replied:
To use multiple LEDs together (e.g., b+g) we need #14029 merged.

SpectreDev1 replied:
yeah thats needed

innovara replied:
On stock firmware, the led flashes slowly after about 10 seconds which is when you are supposed to switch the power off (three times) to switch to the other partition. I am testing your firmware and currently I am counting to do this because it doesn't flash. Is that something that you would consider implementing or is it something that relates to how OpenWrt handles booting up and out of the scope of this PR?

SpectreDev1 replied:
> On stock firmware, the led flashes slowly after about 10 seconds which is when you are supposed to switch the power off (three times) to switch to the other partition. I am testing your firmware and currently I am counting to do this because it doesn't flash. Is that something that you would consider implementing or is it something that relates to how OpenWrt handles booting up and out of the scope of this PR?

on sysupgrade it should automatically switch slots

innovara replied:
Fair enough. I was more thinking of a dual-boot scenario to switch from stock to OpenWrt. I'm not getting the kind of performance that would make OpenWrt permanent just yet, but I certainly hope it will get there.

SpectreDev1 commented:
@Ansuel  All ready to be merged

testuser7 commented on target/linux/qualcommax/ipq807x/base-files/etc/board.d/02_network:
```diff
@@ -39,6 +39,8 @@ ipq807x_setup_interfaces()
 		ucidef_set_interfaces_lan_wan "lan1 lan2 lan3 lan4 10g-2" "10g-1"
 		;;
+        linksys,mx4200v1|\
+        linksys,mx4200v2|\
```
You should keep it in alphabetical order.

innovara replied:
The resolution causes a merge conflict

SpectreDev1 replied:
> The resolution causes a merge conflict

its been like that before fixing it

testuser7 commented on target/linux/qualcommax/ipq807x/base-files/etc/hotplug.d/firmware/11-ath11k-caldata:
```diff
@@ -13,6 +13,8 @@ case "$FIRMWARE" in
 	edgecore,eap102|\
 	edimax,cax1800|\
+	linksys,mx4200v1|\
+	linksys,mx4200v2|\
```
You should keep it in alphabetical order.

innovara replied:
The resolution causes a merge conflict

testuser7 commented on target/linux/qualcommax/ipq807x/base-files/lib/upgrade/platform.sh:
```diff
@@ -66,6 +66,21 @@ platform_do_upgrade() {
 		nand_do_upgrade "$1"
 		;;
+	linksys,mx4200v1|\
+	linksys,mx4200v2)
```
You should keep it in alphabetical order.

innovara replied:
The resolution causes a merge conflict

hauke commented:
Please rebase this pull request on top of current main branch to fix the merge conflicts.

RoninByte committed:
ipq807x: add support for Linksys MX4200 V1 and V2

Linksys MX4200 is a 802.11ax Tri-band router/AP.
Specifications:
* CPU: Qualcomm IPQ8174 Quad core Cortex-A53 1.4GHz
* RAM: 512MB of DDR3
* Storage: 512Mb NAND
* Ethernet: 4x1G RJ45 ports (QCA8075)
* WLAN:
	* 2.4GHz: Qualcomm QCN5024 2x2 802.11b/g/n/ax 574 Mbps PHY rate
	* 5GHz: Qualcomm QCN5054 2x2@80MHz or 2x2@160MHz 802.11a/b/g/n/ac/ax 2402 PHY rate
	* 5GHz: Qualcomm QCN5054 4x4@80MHz or 2x2@160MHz 802.11a/b/g/n/ac/ax 2402 PHY rate
* LED-s:
	* RGB system led

* Buttons: 1x Soft reset 1x WPS
* Power: 12V DC Jack

Installation instructions:
Open Linksys Web UI - http://192.168.1.1/ca or http://10.65.1.1/ca depending on your setup.
Login with your admin password. The default password can be found on a sticker under the device.
To enter into the support mode, click on the “CA” link and the bottom of the page.
Open the “Connectivity” menu and upload the squash-factory image with the “Choose file” button.
Click start. Ignore all the prompts and warnings by click “yes” in all the popups.
The Wifi radios are turned off by default. To configure the router, you will need to connect your computer to the LAN port of the device.
Then you would need to write openwrt to the other partition for it to work
- First Check booted partition
fw_printenv -n boot_part

- Then install Openwrt to the other partition if booted in slot 1:
mtd -r -e alt_kernel -n write openwrt-qualcommax-ipq807x-linksys_mx4200v(X)-squashfs-factory.bin alt_kernel

- If in slot 2:
mtd -r -e kernel -n write openwrt-qualcommax-ipq807x-linksys_mx4200v(X)-squashfs-factory.bin kernel

Replace (X) with your model version either 1 or 2

Signed-off-by: Mohammad Sayful Islam <sayf.mohammad01@gmail.com>
Reviewed-by: Robert Marko <robimarko@gmail.com>

SpectreDev1 commented:
> Please rebase this pull request on top of current main branch to fix the merge conflicts.

All done and sorted

testuser7 commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8174-mx4200.dtsi:
```diff
@@ -0,0 +1,445 @@
+&wifi {
+	status = "okay";
+
+	qcom,ath11k-calibration-variant = "Linksys-MX4200";
```
You should use separate BDF files for v1 and v2 because they are different.

SpectreDev1 replied:
Hmm ok

innovara replied:
For what it's worth, I can confirm that the new BDF file with two variants works for v2

SpectreDev1 replied:
Nice to know it's working for both

testuser7 replied:
> For what it's worth, I can confirm that the new BDF file with two variants works for v2

Check this forum thread: https://forum.openwrt.org/t/qualcommax-ath11k-board-calibration-bdf-data-woes/182646
